modules/react/common/lib/utils/components.ts

```diff
@@ -29,7 +49,7 @@ export type PropsWithAs<
          * If `as` is a component, the reference will be to that component (or element if the component
          * uses `React.forwardRef`).
          */
-        ref?: React.Ref<ElementType>;
+        ref?: ExtractRef<ElementType>;
```

Comment:
Here's the fix. The ref type must be extracted from the `ElementType` generic which could be `undefined`, an `Element`, or an `ElementComponent<'someElementType', SomeProps>`